server/src/main/java/org/opensearch/index/store/remote/utils/cache/SegmentedCache.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.utils.cache;
+
+import org.opensearch.common.cache.RemovalListener;
+import org.opensearch.common.cache.RemovalNotification;
+import org.opensearch.common.cache.Weigher;
+import org.opensearch.index.store.remote.utils.cache.stats.CacheStats;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+import static org.opensearch.common.util.ValidationUtils.checkArgument;
+
+/**
+ * Segmented {@link LRUCache} to offer concurrent access with less contention.
+ * @param <K> type of the key
+ * @param <V> type of th value
+ *
+ * @opensearch.internal
+ */
+public class SegmentedCache<K, V> implements RefCountedCache<K, V> {
```

Comment:
yes it's very similar to ConcurrentHashMap regarding segmentation to reduce contention and because it has very similar interface to a Map but actually its a cache with capacity and eviction and not just a map. Here each segment is LRUCache. I believe this implementation is what Guava Cache is doing at some point (don't know when that code was copied) and it should be compared to Guava/Caffeine Cache. We know that Guava/Caffeine will be more performant than our implementation but we have custom eviction logic (based or reference count) and Guava cache interface is not open for modifying eviction logic. That's why we didn't reused any other code here. In future we might.

Reply:
This works for Java 5's `ConcurrentHashMap` (since rewritten) because entries are distributed uniformly so that writes have a lower chance of collisions (reads are lock-free). However the access distribution for a cache is typically Zipfian (powers law) as hot entries are much more popular than cold ones. In LRU every read is a write, so a lock is required and threads will contend on the hot entries. This is why segmenting an LRU cache has only marginal benefits, but that may still be fine in many usages (e.g. memcached simply does not lock+reorder if the entry has been touched within the last 60s).

Reply:
@ben-manes so are you saying even the Java 5 version of `ConcurrentHashMap` was sub-optimal due to the access pattern requirements of LRU thus we need our own implementation?

Reply:
I believe so. `ConcurrentHashMap` does not provide LRU bounding and does not mutate any state on a read. This allows it to be lock-free, so read access patterns do not matter for its performance. It segments because it knows that entries are uniformly distributed on write so this allows for using more lock to reduce contention. As reads are the common case and basically free, coarse grained locking on writes was fine. In the v8 that introduced computations, the redesign uses fine-grained segments (hash bins) because the lock may be held for long durations by user code, so excessive blocking became a problem again.

An traditional LRU requires that every read mutates shared state to maintain access order. As in your case, this is done using a doubly-linked list for O(1) reorderings, where items are moved from the middle to the end (tail) and eviction polls from the front (head). That complexity too much to do using CAS alone and would be inefficient (the tail would be a contended on). While the operations are cheap, the locking serializes all usages and can become a bottleneck. This is made worse by your computations being under this shared lock, rather than for example storing a future that is completed outside of the lock (see JCiP). In general you can expect ~10M ops/s for read/write operations (see `LinkedHashMap` in these [benchmarks](https://github.com/ben-manes/caffeine/wiki/Benchmarks#server-class)).

An LRU is access-order and it guesses that this will provide the highest hit rate. That means you have a skewed distribution - uniform would be random replacement. Since you know hot >> cold, certain entries are queried for much more frequently than others. Those are guarded by an LRU lock, so hot keys will cause contention. While you may see some minor benefit by spreading hot entries across different locks, their popularity means you will won't get a large speedup (let's say 2.5x). If that's not fast enough, then you have to use other tricks (like memcached's age threshold) to get to acceptable performance. All of this means is that the problems are harder than they seem and the traditional approaches predated highly parallel, prevalent multi-core usages.

As noted elsewhere, it looks like you could plug Caffeine into your proposed `incRef` / `decRef` to pin the entry. That would simply modify a counter on your wrapped value within a `Map.compute`, and the weigher take that into account in its evaluation.

```java
Cache<K, V> cache = Caffeine.newBuilder()
    .maximumWeight(10_000)
    .weigher((K key, V value) -> value.evictable() ? value.weight() : 0)
    .build();

void incRef(K key) {
  cache.asMap().computeIfPresent(key, value -> {
      value.refCount++;
      return value;
  });
}
void decRef(K key) {
  cache.asMap().computeIfPresent(key, value -> {
      value.refCount--;
      return value;
  });
}
```